### ISSUE TITLE
Intermittent failure of twisted.names.test.test_names.ServerDNSTestCase.testZoneTransfer on OS X

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3424]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3424|
|Type|defect|
|Created|2008-09-08 23:00:18Z|
|Branch|https://github.com/twisted/twisted/tree/dns-test-cleanup-3424|

```
[ERROR]: twisted.names.test.test_names.ServerDNSTestCase.testZoneTransfer

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<DNSProtocol #0 on 58948>
```

<details><summary>Searchable metadata</summary>

```
trac-id__3424 3424
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_dns_test_cleanup_3424 branches/dns-test-cleanup-3424
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1220914818000000 1220914818000000
changetime__1221252519000000 1221252519000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This is because the server side of the TCP connection established to perform the zone transfer isn't waited for when the test is finished.  In fact, nothing about it is exposed through any interface, so it would be hard to wait for it to disconnect.  `DNSServerFactory` needs a new method, I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [24788]) Branching to 'dns-test-cleanup-3424'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Should be fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

One unrelated flake: twisted/names/test/test_dns.py:506: redefinition of function 'test_dname' from line 496

Everything else is fine, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [24813]) Merge dns-test-cleanup-3424

Author: exarkun
Reviewer: therve
Fixes: [#3424](https://github.com/twisted/twisted/issues/3424)

Change `DNSServerFactory` so that it keeps track of all the TCP
clients connected to it.  Use this information in the tests to
make sure all selectables are cleaned up before the test ends
so that trial does not report that the reactor is unclean.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
